Add support for "Type A: Activity Manged" activities

mmoore99 opened the pull request:
Fixes #506

**This PR should be able to be merged into "Master" as it does not  introduce any breaking changes.  Unless an Activity is specifically assigned an "ActivityType" of "ActivityManaged" the original workflow for avtivities and tasks remains intact.**

**Note: this PR depends on PR #514**

This PR implements the following changes to the functionality of the Activity Detail Page:

-  No task information is displayed for an "Activity Managed" activity
- The "Volunteer" link was changed to a button and made more prominent (**see Screenshot 1**)
- Client side validation has been implemented for the activity signup dialog using Knockout (**see Screenshot 2**)
- A confirmation message is now displayed after a successfull signup (**see Screenshot 3**)
- A confirmation message is now displayed after unenrolling from an activity (**see Screenshot 4**)
- The posting of the signup activity dialog is now done through an Ajax call to the ActivityApi conntroller, thereby eliminating the need for a page refresh after signing up.  While, this may not be a specific requirement for this issue, it is needed to support functionality for issue #507 where the user will be picking from a list of tasks on the page.

This PR also fixes a bug in the existing code when a user unenrolls from an Activity where task(s) have been assigned to the user. The code was attempting to delete "Tasks" instead of "TaskSignups" and also was causing system exception to be thrown.

**Screenshot 1**
![screenclip 1](https://cloud.githubusercontent.com/assets/45022/12985571/339b0c38-d0a8-11e5-8b53-7655495bba75.png)

**Screenshot 2**
![screenclip 2](https://cloud.githubusercontent.com/assets/45022/12985620/75d715f6-d0a8-11e5-8cba-70ef7ed1ba14.png)

**Screenshot 3**
![screenclip 3](https://cloud.githubusercontent.com/assets/45022/12985627/7d256ed4-d0a8-11e5-90df-fa7aa238d4e8.png)

**Screenshot 4**
![screenclip 4](https://cloud.githubusercontent.com/assets/45022/12985635/83963cbc-d0a8-11e5-9d82-097e9bdfb30c.png)

MisterJames commented:
We'll have to sort out the build failure before merging this one. I'll start investigating that now.

MisterJames commented:
@mmoore99 It looks like the test that is failing is in the neighbourhood of activity signup. Could you please evaluate this further?

mmoore99 commented:
@MisterJames I have investigated this  a little further.  It appears that the test is failing while trying to confirm the ability for the admin to **create an activity** ("Test: The Org Admin can create activies").  Nothing in this PR involves that process.  

However, this PR depends on PR #514, which has not yet been merged.  #514 does make some changes to the Activity create process as it has added the "ActivityType" field.  So, possibly, this error will go away once #514 is merged?  Just sort of guessing here since I am not really familiar with the scenario test process.

mmoore99 commented:
@MisterJames upon further investigation, I think that "Test: The Org Admin can create activities"  may be failing since in #514 there was some validation added to the "Add Activity" form.  The "ActivityType" property was added and is "Required".  Also, the "NumberOfVolunteersRequired " was made "Required". Since the scenario test does not populate either of these fields, the form submission results in validation errors instead of a successful Activity creation.

Is it possible to run the scenario tests locally?

mmoore99 commented:
@MisterJames I pushed a change to #514 in an attempt to fix the scenario test.  Please see comment under that PR.

tonysurma commented:
along with the comment on 514 I would like to merge this after that but am not sure if the test error - anyone have ideas how to resolve? thanks.

tonysurma commented:
same here @mheggeseth @BillWagner @MisterJames will merge this after we get a verdict on the other and the fail here.  THanks

mheggeseth commented:
Looks like this could still use some code review. I'm happy to do it.

mmoore99 commented:
@mheggeseth do you want me to do a rebase on this since the commits from #514 are still showing in this PR?

mheggeseth commented:
@mmoore99 Sure, go ahead and rebase.

mmoore99 commented:
@mheggeseth rebase completed.

mheggeseth commented on AllReadyApp/Web-App/AllReady/Views/Activity/Activity.cshtml:
```diff
@@ -120,34 +122,35 @@
+                    if (!Model.IsClosed)
+                     {
+                         <h2>
+                             <button class="btn btn-primary btn-volunteer" data-bind="visible: !enrolled(), click: signupForActivity" style="display: none">Volunteer<br /> for this Activity</button>
```
The Volunteer button shouldn't look different depending on whether you are signed in. I think just saying "Volunteer" is sufficient. I would also make the button orange to be consistent with the rest of the action buttons on the site.

Signed in:
![image](https://cloud.githubusercontent.com/assets/775229/13198980/01b99ad6-d7dd-11e5-87a2-97238f903d76.png)


Not signed in:
![image](https://cloud.githubusercontent.com/assets/775229/13198971/daa452ba-d7dc-11e5-81f1-1e9fafcf6df8.png)

mmoore99 replied:
@mheggeseth I'll change the color as suggested. I agree that the button should be the same whether signed in or not.  Regarding the wording, I made the change in wording specifically in response to @tonysurma's request in #502:

**`Volunteer link is present and worded with clarity not just 'volunteer' but something like 'Volunteer for this Campaign'`**

I'll be glad to change it to whatever is decided.  I'll let you make the call.

mheggeseth replied:
OK, you can change it back to "Volunteer for this activity"

mheggeseth commented on AllReadyApp/Web-App/AllReady/Views/Activity/Activity.cshtml:
```diff
@@ -292,13 +301,14 @@
-    </script>
+        </script>
+    <script type="text/javascript" src="~/js/ko.bindings.js"></script>
+    <script type="text/javascript" src="~/lib/knockout.mapping/knockout.mapping.js"></script>
```
Looks like you forgot to commit the bower.json changes that add `knockout.mapping`.

mmoore99 replied:
@mheggeseth the reason for this is that when I went to bower to try to find the  knockout-mapping plugin I found three entries, all of which were forks of Steve Sanderson's "official" plugin.  The three that I found are [here](https://github.com/crissdev/knockout.mapping), [here](https://github.com/LRPalacios/knockout.mapping) and [here](https://github.com/crissdev/knockout.mapping).  Steve Sanderson's version is [here](https://github.com/SteveSanderson/knockout.mapping).  So, instead of using one of the forks i just downloaded Sanderson's.

Do you want me to swtch to one of the forks or just continue with the downloaded version?

mheggeseth replied:
You can keep using the downloaded version. You will just have to explicitly add knockout.mapping.js` to the git repo because `.gitignore` is set up to ignore everything in `~/lib`.

mmoore99 commented:
@mheggeseth I committed changes for the color and wording of volunteer button as suggested.  If the wording on the button needs to be changed to something else, let me know.

mheggeseth commented on AllReadyApp/Web-App/AllReady/Startup.cs:
```diff
@@ -19,6 +19,7 @@
 using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.PlatformAbstractions;
 using AllReady.Security;
+using Newtonsoft.Json.Serialization;
```
Don't think this is necessary.

mmoore99 commented:
@mheggeseth I pushed changes to remove unused `using` in startup.cs and added `ko.mapping` files to source control

mheggeseth commented:
Fix the punctuation here. The "...signed up for this activity" sentence doesn't end and "We hope to see you soon" ends twice.
![image](https://cloud.githubusercontent.com/assets/775229/13199428/b505ecae-d7ea-11e5-82ed-00b2de43062e.png)

mmoore99 commented on AllReadyApp/Web-App/AllReady/Views/Activity/Activity.cshtml:
```diff
@@ -7,6 +7,8 @@
     var userTimeZoneId = User.GetTimeZoneId();
 }
 
+<div id="alert_placeholder"></div>
```
@mheggeseth Yes, I agree. I had thought about that at one time, but forgot to come back to it.  I'll work on some changes.

mheggeseth commented on AllReadyApp/Web-App/AllReady/Controllers/ActivityApiController.cs:
```diff
@@ -140,33 +142,24 @@ public ActionResult GetCheckin(int id)
-        public async Task<IActionResult> RegisterActivity(int id)
+        [HttpPost("signup")]
+        [Authorize]
+        public async Task<IActionResult> RegisterActivity(ActivitySignupViewModel signupModel)
```
This doesn't need to be async because you aren't awaiting anything in the body.

mheggeseth commented on AllReadyApp/Web-App/AllReady/DataAccess/AllReadyDataAccessEF7.ActivitySignup.cs:
```diff
@@ -39,24 +39,31 @@ Task IAllReadyDataAccess.AddActivitySignupAsync(ActivitySignup userSignup)
             var activity = _dbContext.ActivitySignup.SingleOrDefault(c => c.Id == activitySignupId);
 
-            if (activity != null)
+            if (activity == null) return null;
```
Although technically correct, the way the code is formatted makes your intentions somewhat unclear ("do you mean to return null when activity is null or execute the block of code?"). Something like this would be better:

``` CSharp
if (activity == null)
{
  return null; //Purposely on its own line to draw attention to the fact that this is an exit point
}

//The real work...
```

mmoore99 replied:
@mheggeseth agreed!

mheggeseth commented on AllReadyApp/Web-App/AllReady/DataAccess/AllReadyDataAccessEF7.ActivitySignup.cs:
```diff
@@ -39,24 +39,31 @@ Task IAllReadyDataAccess.AddActivitySignupAsync(ActivitySignup userSignup)
-        Task IAllReadyDataAccess.DeleteActivitySignupAsync(int activitySignupId)
+        Task IAllReadyDataAccess.DeleteActivityAndTaskSignupsAsync(int activitySignupId)
         {
             var activity = _dbContext.ActivitySignup.SingleOrDefault(c => c.Id == activitySignupId);
```
Local variable should probably be called `activitySignup` instead of `activity` since that's what it is.

mmoore99 replied:
@mheggeseth Agreed.  I had noticed this as well when I was working on PR #523 and made the change there.  I'll make the change here as well.

mheggeseth commented on AllReadyApp/Web-App/AllReady/DataAccess/AllReadyDataAccessEF7.ActivitySignup.cs:
```diff
@@ -39,24 +39,31 @@ Task IAllReadyDataAccess.AddActivitySignupAsync(ActivitySignup userSignup)
+                    var taskSignup = _dbContext.TaskSignups.SingleOrDefault(c => c.Id == signupId);
+                    if (taskSignup != null)
+                    {
+                        _dbContext.TaskSignups.Remove(taskSignup);
```
Instead of getting the TaskSignups, getting their IDs, then getting the TaskSignups again to delete them, can't you do something like: 
```
_dbContext.TaskSignups.RemoveRange(_dbContext.TaskSignups
    .Where(e => e.Task.Activity.Id == activity.Activity.Id)
    .Where(e => e.User.Id == activity.User.Id)
);
```

mmoore99 replied:
@mheggeseth yes, that is much cleaner.  Thanks.

mheggeseth commented on AllReadyApp/Web-App/AllReady/wwwroot/js/activity.js:
```diff
@@ -1,57 +1,118 @@
-                console.log(fail);
+        function activitySignupSuccess(signUpViewModel) {
+            self.skills().forEach(function(skill) {
+                if (signUpViewModel.AddSkillIds().indexOf(skill.Id) >= 0){ skill.IsUserSkill(true) }
```
I can't find any evidence in the code where `AddSkillIds` becomes an observable array. I think it's just a plain array. Additionally, you should put the curly braces on separate lines and put a semicolon after `skill.IsUserSkill(true)`.

mmoore99 replied:
@mheggeseth `AddSkillIds` is an observable array.  It is created in the ko.mapping statement when the view model is created.  Since AddSkillIds is passed down from the server as an array, the ko.mapping makes it observable.
```
SignupViewModel = function (signupModelSeed, unassociatedSkills) {
        var self = this;
        ko.mapping.fromJS(signupModelSeed, {}, self);
        self.unassociatedSkills = unassociatedSkills;

I'll make the formatting changes as suggested.

mheggeseth commented on AllReadyApp/Web-App/AllReady/wwwroot/js/ko.bindings.js:
```diff
@@ -0,0 +1,20 @@
+    },
+    update: function (element, valueAccessor) {
+        var value = ko.utils.unwrapObservable(valueAccessor());
+        $(element).val(value);
```
So I like that you wrapped the `$().mask` plugin with a binding handler. However, the way you've written it you kind of have to implement your own `value` binding handler at the same time. If you search `knockout.debug.js` for `ko.bindingHandlers['value']` you'll see that you probably don't want to do that. I would instead suggest trying to get the mask plugin to play nicely with KO's standard value binding (if you haven't already tried that). There are a couple ways you could do that:
* `masked` could wrap the `value` binding and call through to `ko.bindingHandlers.value.init/update`. In this case I would suggest naming the binding `valueMasked` to more clearly convey that it is derivative of `value`
* `masked` could be its own binding independent of `value` and try not deal with the value of the `input` at all

Additionally, when you have a binding handler that takes multiple options, I think it's better practice for the bound value to be an object containing each of the options, rather than conveying those options through sibling bindings via allBindingsAccessor`: `masked: { placeholder: "", mask: "" }`

mheggeseth commented on AllReadyApp/Web-App/AllReady/wwwroot/js/activity.js:
```diff
@@ -1,57 +1,118 @@
+        };
+    };
+
+    SignupViewModel = function (signupModelSeed, unassociatedSkills) {
```
This function will live in the global namespace unless you do one of these things:
* `var SignupViewModel = function (signupModelSeed, unassociatedSkills)`
*  `function SignupViewModel(signupModelSeed, unassociatedSkills)`

mheggeseth commented on AllReadyApp/Web-App/AllReady/wwwroot/js/activity.js:
```diff
@@ -1,57 +1,118 @@
+        self.isValid = ko.computed(function () {
+            var allValidatablesAreValid = true;
+            for (var property in self) {
+                if (self.hasOwnProperty(property) && self[property]["hasError"]) {
```
Use `typeof self[property].hasError === "function"`

mheggeseth commented on AllReadyApp/Web-App/AllReady/wwwroot/js/ko.validation.js:
```diff
@@ -17,4 +17,56 @@ ko.observableArray.fn.uniqueValidator = function (property) {
+// This is a "poor man's" validation routine that provides specific, limited capabilities (currently only supports required, email, phoneNumber validations).
+// If more a more general & extensive validation capability is needed, consider using Knockout-Validation plugin at https://github.com/Knockout-Contrib/Knockout-Validation
+//
+ko.extenders.validate = function (target, options) {
```
I like what you did here, but the syntax for using `validate` is a bit clunky. For instance, it's not very obvious what it means to set `phoneNumber` equal to the empty string. It kind of forces you to go read the implementation of `validate`. Additionally, the maintainability of the `validate` function is somewhat suspect. Do you just keep adding if blocks to it for each kind of validation you want to implement? And each of the if blocks has basically the same format and should probably be refactored into a single function.
```
self.PreferredPhoneNumber.extend({
    validate: {
        required: "'Phone Number' is required",
        phoneNumber: ""
    }
});
```

I think it might work better to use the `withValidator` fn in a way that you could then chain validation functions together something like
```
self.PreferredPhoneNumber
    .isRequired({ displayName: "Phone Number"})
    .validatePhoneNumber()
```

`withValidator` would need to be enhanced to be chainable. This isn't too bad. You just check whether `isValid` is already defined and combine its result with the result of `validatorFn` when you re-define `isValid` as a new computed observable. Notice I also changed the base type to be `subscribable` so you can use it with arrays or single values. This also achieves the same short-circuiting that your validate function does because `validateFn` will not execute unless `isValidAlready` returns true.
```
ko.subscribable.fn.withValidator = function (validatorFn) {
    var isValidAlready = this.isValid || function () { return true; } //Default to valid if no validator defined yet
    this.isValid = ko.computed(function () {
        return isValidAlready() && validatorFn(this);
    }, this);
    return this;
};
```

You could also pretty easily add an option to defer validation until the observable changes from its initial value. That seems basically what you are getting at with `showErrors`. If you want to work together on this, I'd be happy to.

mheggeseth commented on AllReadyApp/Web-App/AllReady/Views/Shared/_Layout.cshtml:
```diff
@@ -136,7 +136,6 @@
         </script>
     </environment>
     <script src="~/lib/knockout-bootstrap/src/knockout-bootstrap.js"></script>
-    <script src="~/lib/system.js/dist/system.js"></script>
```
system.js is actually being used on Views\Campaign\Details.cshtml

mheggeseth commented:
I'm all for using AJAX for the volunteer workflow, but some things about the way it's implemented are perhaps a little more complicated than they need to be. I'm not sure the whole dynamic template rendering thing in `showModal` is necessary. I'm actually not sure `showModal` itself is necessary (or that its benefits are worth its added complexity). Another thing is that the "form" doesn't actually need to be a form anymore because it no longer has a submit action. Do you want to pair on this and see what we can simplify or get rid of?

mmoore99 commented:
@mheggeseth sure, I would be glad to pair with you on this.  I debated as well about whether the `showModal` approach was necessary and would like to discuss it with you.  I am going to be tied up for the rest of the evening, but would have time tomorrow, if that works for you.  Let me know.

mheggeseth commented:
I'm available tomorrow morning starting at 9 central.

mheggeseth commented on AllReadyApp/Web-App/AllReady/Controllers/ActivityApiController.cs:
```diff
@@ -140,34 +142,25 @@ public ActionResult GetCheckin(int id)
-            await _allReadyDataAccess.UpdateActivity(activity);
-            return new HttpStatusCodeResult((int)HttpStatusCode.OK);
+            _bus.Send(new ActivitySignupCommand() { ActivitySignup = signupModel });
+            return new {Status = "success"};
```
Despite the fact that it's not async anymore, I think you should still return an `IActionResult`. Just `return Json(null)` if everything went great.

mmoore99 commented:
@mheggeseth How about 10 central?  I'll Skype you.

mheggeseth commented:
Good

tonysurma commented:
@mmoore99 and @mheggeseth did you get a chance to chat today?  Thanks.

mmoore99 commented:
@tonysurma @mheggeseth  Yes, we talked today.  I've got some more work to do before we'll be ready to merge.  I'll update when ready.

tonysurma commented:
Thank you!  You really have been contributing an amazing amount @mmoore99

mmoore99 commented:
@tonysurma You're welcome, I'm having fun.  Unlike your other contributors, I think, I'm an old retired guy that likes to write code.  Keeps the brain functioning :sunglasses:

mmoore99 commented:
@mheggeseth I pushed the changes for the refactored `showModal`

mmoore99 commented:
- @mheggeseth, cc:@tonysurma  I have pushed changes that I believe address all of the outstanding items that we have been discussing on this PR. 

- A couple of notes:
  - On the knockout custom binding for the jquery edit mask I removed the `placeholder` option as it is not needed since we can just use the normal html `placeholder` attribute.

  - @BillWagner  @SeanFeldman I rebased on top of the PR's that were merged last weekend, including the async stuff.  In `ActivitySignupHandler.cs` the call to save database changes was changed to `await _context.SaveChangesAsync();`. When testing, this statement throws the exception below.  I changed the code back to `_context.SaveChanges();` and it does not throw the exception.  So I am going to leave it this way for now until we can determine what is going on.

> Microsoft.Data.Entity.DbContext: Error: An exception occurred in the database while saving changes.
> Microsoft.Data.Entity.DbUpdateException: An error occurred while updating the entries. See the inner exception for details. ---> System.InvalidOperationException: Invalid attempt to call FieldCount when reader is closed.
>    at System.Data.SqlClient.SqlDataReader.get_FieldCount()
>    at Microsoft.Data.Entity.Storage.Internal.UntypedRelationalValueBufferFactory.Create(DbDataReader dataReader)
>    at Microsoft.Data.Entity.Update.AffectedCountModificationCommandBatch.<ConsumeResultSetWithPropagationAsync>d__8.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at Microsoft.Data.Entity.Update.AffectedCountModificationCommandBatch.<ConsumeAsync>d__6.MoveNext()
>    --- End of inner exception stack trace ---
>    at Microsoft.Data.Entity.Update.AffectedCountModificationCommandBatch.<ConsumeAsync>d__6.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
>    at Microsoft.Data.Entity.Update.ReaderModificationCommandBatch.<ExecuteAsync>d__28.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
>    at Microsoft.Data.Entity.Update.Internal.BatchExecutor.<ExecuteAsync>d__1.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at Microsoft.Data.Entity.ChangeTracking.Internal.StateManager.<SaveChangesAsync>d__32.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at Microsoft.Data.Entity.DbContext.<SaveChangesAsync>d__28.MoveNext()
> Exception thrown: 'Microsoft.Data.Entity.DbUpdateException' in mscorlib.dll
>

mmoore99 commented:
pushed changes that refactor `showModal` to have fluent api and better encapsulate modal hide function

SeanFeldman commented:
@mmoore99 not sure what's the issue. `ActivitySignupHandler` doesn't have a test :disappointed: so I manually ran through the scenario (volunteer enrollment into an activity) and it works ok. Is that exception happening all the time for you?

SeanFeldman commented:
Though I did miss one thing for sure:
```c#
await _context.SaveChangesAsync();
```
should be
```c#
await _context.SaveChangesAsync().ConfigureAwait(false);
```

mmoore99 commented:
@SeanFeldman yes, it is happening all of the time when the async call is used. I tried it using `await _context.SaveChangesAsync().ConfigureAwait(false);` with the same result.  I also tried it with a new database, same result.  It works fine with the non async call.  Of course, I am running on a different code base than you are.  I have this PR plus another one,  #523 , which is coming right behind this one.  I would propose that we leave the non async call in until we get past these two PR's and then we can see where we're at.

SeanFeldman commented:
@mmoore99 sounds good. The fact that there's no unit test to run against makes it a bit more difficult.

tonysurma commented:
@mmoore99   please rebase off of master as we now have a fix for the build/test issues!

For reference on rebasing this page helps (replace upstream with HTBox in examples) https://robots.thoughtbot.com/keeping-a-github-fork-updated

mmoore99 commented:
@tonysurma rebase on master complete

tonysurma commented:
@mheggeseth just dbl checking you are good to go on this and I will merge.  @mmoore99 thanks for the rebase

mheggeseth commented:
I haven't read all of @mmoore99's new changes but we've been in close contact so I good with this.

SeanFeldman commented:
@mheggeseth @mmoore99 is the async change got reverted back to be there or still requires attention.

mheggeseth commented:
@SeanFeldman The proposed code in `ActivitySignupHandler` is as you and @mmoore99 agreed upon yesterday: `_context.SaveChanges()` instead of `await _context.SaveChangesAsync()`.

However, I noticed something peculiar. I wonder whether the real problem here is that the sending of `ActivitySignupCommand` via `_bus.SendAsync()` isn't being "awaited" in `ActivityApiController.RegisterActivity`. I realize you probably _should_ be able to use Mediatr in this way, but I'd be interested to know if it works to change back to `await _context.SaveChangesAsync()` if you also `await _bus.SendAsync(new ActivitySignupCommand())`

SeanFeldman commented:
Asynchronous operation should always be awaited, unless the task is
returned directly to the caller. Could be the issue indeed.
On Feb 26, 2016 19:29, "Mike Heggeseth" <notifications@github.com> wrote:

> @SeanFeldman <https://github.com/SeanFeldman> The proposed code in
> ActivitySignupHandler is as you and @mmoore99
> <https://github.com/mmoore99> agreed upon yesterday:
> _context.SaveChanges() instead of await _context.SaveChangesAsync().
>
> However, I noticed something peculiar. I wonder whether the real problem
> here is that the sending of ActivitySignupCommand via _bus.SendAsync()
> isn't being "awaited" in ActivityApiController.RegisterActivity. I
> realize you probably *should* be able to use Mediatr in this way, but I'd
> be interested to know if it works to change back to await
> _context.SaveChangesAsync() if you also await _bus.SendAsync(new
> ActivitySignupCommand())
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HTBox/allReady/pull/518#issuecomment-189559688>.
>

mmoore99 commented:
@mheggeseth @SeanFeldman.  It appears that @mheggeseth's comments regarding the lack of "await" on `_bus.SendAsync()`  in `ActivityApiController.RegisterActivity`being the problem are correct.  I put the "await" on that call and changed the call back to `await _context.SaveChangesAsync()` `ActivitySignupHandler` and everything worked properly.  Great catch @mheggeseth!  Since this PR has not been merged yet, I'll go ahead and push up these changes.

SeanFeldman commented:
:+1:

mheggeseth commented:
:shipit:

mmoore99 commented:
@BillWagner @tonysurma Now that the review of this PR is complete, would it be possible to merge it? Both this PR and #523 (which is under active development) are in a lot of the same code areas and keeping the two branches alive and updated is time consuming.  Thanks.
